src/predictor.rs

```diff
@@ -0,0 +1,659 @@
+//! Predictors for no predictor, horizontal and floating-point
+use std::collections::HashMap;
+use std::fmt::Debug;
+
+use bytes::{Bytes, BytesMut};
+// use tiff::decoder::DecodingResult;
+
+use crate::{
+    error::AsyncTiffResult, reader::Endianness, tiff::tags::Predictor, tile::PredictorInfo,
+};
+
+/// A registry for reverse predictors
+///
+/// Reverse predictors, because they perform the inverse (decoding) operation of prediction
+///
+///
+///
+pub struct RevPredictorRegistry(HashMap<Predictor, Box<dyn RevPredict>>);
+
+impl RevPredictorRegistry {
+    /// create a new predictor registry with no predictors registered
+    pub fn new() -> Self {
+        Self(HashMap::new())
+    }
+}
+
+impl AsRef<HashMap<Predictor, Box<dyn RevPredict>>> for RevPredictorRegistry {
+    fn as_ref(&self) -> &HashMap<Predictor, Box<dyn RevPredict>> {
+        &self.0
+    }
+}
+
+impl Default for RevPredictorRegistry {
+    fn default() -> Self {
+        let mut hmap = HashMap::new();
+        hmap.insert(Predictor::None, Box::new(NoPredictor) as _);
+        hmap.insert(Predictor::Horizontal, Box::new(RevHorizontalPredictor) as _);
+        hmap.insert(
+            Predictor::FloatingPoint,
+            Box::new(RevFloatingPointPredictor) as _,
+        );
+        Self(hmap)
+    }
+}
```

Comment:
As mentioned previously, I think simplest to remove the registry concept from here. Then in `Tile::decode`, just match on the `Predictor` tag enum manually.